File: test/framework/harnesses/host_controller/tradefed/remote_client.py
#
# Copyright (C) 2017 The Android Open Source Project
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

import logging
import socket
import time

from host_controller.tradefed import remote_operation

LOCALHOST = "localhost"
DEFAULT_PORT = 30103


class RemoteClient(object):
    """The class for sending remote operations to TradeFed."""

    def __init__(self, host=LOCALHOST, port=DEFAULT_PORT, timeout=None):
        """Initializes the client of TradeFed remote manager.

        Args:
            _host: The host name of the remote manager.
            _port: The port of the remote manager.
            _timeout: The connect and receive timeout in seconds
        """
        self._host = host
        self._port = port
        self._timeout = timeout if timeout else socket.getdefaulttimeout()

    def SendOperations(self, *operations):
        """Sends a list of operations and waits for each response.

        Args:
            *operations: A list of remote_operation.RemoteOperation objects.

        Returns:
            A list of JSON objects.

        Raises:
            socket.error if fails to communicate with remote manager.
            remote_operation.RemoteOperationException if any operation fails or
            has no response.
        """
        op_socket = socket.create_connection((self._host, self._port),
                                             self._timeout)
        logging.info("Connect to %s:%d", self._host, self._port)
        try:
            if self._timeout is not None:
                op_socket.settimeout(self._timeout)
            ops_str = "\n".join(str(op) for op in operations)
            logging.info("Send: %s", ops_str)
            op_socket.send(ops_str)
            op_socket.shutdown(socket.SHUT_WR)
            resp_str = ""
            while True:
                buf = op_socket.recv(4096)
                if not buf:
                    break
                resp_str += buf
        finally:
            op_socket.close()
        logging.info("Receive: %s", resp_str)
        resp_lines = [x for x in resp_str.split("\n") if x]
        if len(resp_lines) != len(operations):
            raise remote_operation.RemoteOperationException(
                    "Unexpected number of responses: %d" % len(resp_lines))
        return [operations[i].ParseResponse(resp_lines[i])
                for i in range(len(resp_lines))]

    def SendOperation(self, operation):
        """Sends one operation and waits for its response.

        Args:
            operation: A remote_operation.RemoteOperation object.

        Returns:
            A JSON object.

        Raises:
            socket.error if fails to communicate with remote manager.
            remote_operation.RemoteOperationException if the operation fails or
            has no response.
        """
        return self.SendOperations(operation)[0]

    def ListDevices(self):
        """Sends ListDevices operation.

        Returns:
            A list of device_info.DeviceInfo which are the devices connected to
            the host.
        """
        json_obj = self.SendOperation(remote_operation.ListDevices())
        return remote_operation.ParseListDevicesResponse(json_obj)

    def WaitForCommandResult(self, serial, timeout, poll_interval=5):
        """Sends a series of operations to wait until a command finishes.

        Args:
            serial: The serial number of the device.
            timeout: A float, the timeout in seconds.
            poll_interval: A float, the interval of each GetLastCommandResult
                           operation in seconds.

        Returns:
            A JSON object which is the result of the command.
            None if timeout.

            Sample
            {'status': 'INVOCATION_SUCCESS',
             'free_device_state': 'AVAILABLE'}
        """
        deadline = time.time() + timeout
        get_result_op = remote_operation.GetLastCommandResult(serial)
        while True:
            result = self.SendOperation(get_result_op)
            if result["status"] != "EXECUTING":
                return result
            if time.time() > deadline:
                return None
            time.sleep(poll_interval)
            if time.time() > deadline:
                return None

File: tools/tradefederation/core/remote/src/com/android/tradefed/command/remote/RemoteClient.java
/*
 * Copyright (C) 2011 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.tradefed.command.remote;

import com.android.ddmlib.Log;

import org.json.JSONException;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.net.InetAddress;
import java.net.Socket;
import java.net.UnknownHostException;
import java.util.List;

/**
 * Class for sending remote commands to another TF process.
 * <p/>
 * Currently uses JSON-encoded data sent via sockets.
 */
public class RemoteClient implements IRemoteClient {

    // choose an arbitrary default port that according to the interweb is not used by another
    // popular program
    public static final int DEFAULT_PORT = 30103;

    private static final String TAG = RemoteClient.class.getSimpleName();
    private final Socket mSocket;
    private final PrintWriter mWriter;
    private final BufferedReader mReader;

    /**
     * Initialize the {@RemoteClient}, and instruct it to connect to the given port
     * on localhost.
     *
     * @param port the tcp/ip port number
     * @throws IOException
     * @throws UnknownHostException
     */
    private RemoteClient(int port) throws UnknownHostException, IOException {
        this(InetAddress.getLocalHost().getHostName(), port);
    }

    /**
     * Initialize the {@RemoteClient}, and instruct it to connect to the given hostname and port.
     *
     * @param hostName to connect to
     * @param port the tcp/ip port number
     * @throws IOException
     * @throws UnknownHostException
     */
    private RemoteClient(String hostName, int port) throws UnknownHostException, IOException {
        mSocket = new Socket(hostName, port);
        mWriter = new PrintWriter(mSocket.getOutputStream(), true);
        mReader = new BufferedReader(new InputStreamReader(mSocket.getInputStream()));
    }

    /**
     * Send the given operation to the remote TF.
     *
     * @param op the {@link RemoteOperation} to send
     * @throws RemoteException if failed to perform operation
     */
    private synchronized <T> T sendOperation(RemoteOperation<T> op) throws RemoteException {
       try {
           Log.d(TAG, String.format("Sending remote op %s", op.getType()));
           mWriter.println(op.pack());
           String response = mReader.readLine();
           if (response == null) {
               throw new RemoteException("no response from remote manager");
           }
           return op.unpackResponseFromString(response);
       } catch (IOException e) {
           throw new RemoteException(e.getMessage(), e);
       } catch (JSONException e) {
           throw new RemoteException(e.getMessage(), e);
       }
    }

    /**
     * Helper method to create a {@link RemoteClient} connected to given port
     *
     * @param port the tcp/ip port
     * @return the {@link RemoteClient}
     * @throws RemoteException if failed to connect
     */
    public static IRemoteClient connect(int port) throws RemoteException {
        try {
            return new RemoteClient(port);
        } catch (IOException e) {
            throw new RemoteException(e);
        }
    }

    /**
     * Helper method to create a {@link RemoteClient} connected to given host and port
     *
     * @param hostname the host name
     * @param port the tcp/ip port
     * @return the {@link RemoteClient}
     * @throws RemoteException if failed to connect
     */
    public static IRemoteClient connect(String hostname, int port) throws RemoteException {
        try {
            return new RemoteClient(hostname, port);
        } catch (IOException e) {
            throw new RemoteException(e);
        }
    }

    /**
     * Helper method to create a {@link RemoteClient} connected to default port
     *
     * @return the {@link RemoteClient}
     * @throws RemoteException if failed to connect
     */
    public static IRemoteClient connect() throws RemoteException {
        return connect(DEFAULT_PORT);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void sendAllocateDevice(String serial) throws RemoteException {
        sendOperation(new AllocateDeviceOp(serial));
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void sendFreeDevice(String serial) throws RemoteException {
        sendOperation(new FreeDeviceOp(serial));
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void sendAddCommand(long totalTime, String... commandArgs) throws RemoteException {
        sendOperation(new AddCommandOp(totalTime, commandArgs));
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void sendAddCommandFile(String commandFile, List<String> extraArgs)
            throws RemoteException {
        sendOperation(new AddCommandFileOp(commandFile, extraArgs));
    }

    /**
     * {@inheritDoc}
     */
    @Deprecated
    @Override
    public void sendClose() throws RemoteException {
        sendOperation(new CloseOp());
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void sendStartHandover(int port) throws RemoteException {
        sendOperation(new StartHandoverOp(port));
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void sendHandoverInitComplete() throws RemoteException {
        sendOperation(new HandoverInitCompleteOp());
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void sendHandoverComplete() throws RemoteException {
        sendOperation(new HandoverCompleteOp());
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public List<DeviceDescriptor> sendListDevices() throws RemoteException {
        return sendOperation(new ListDevicesOp());
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void sendExecCommand(String serial, String[] commandArgs) throws RemoteException {
        sendOperation(new ExecCommandOp(serial, commandArgs));
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void sendGetLastCommandResult(String serial, ICommandResultHandler handler)
            throws RemoteException {
        CommandResult r = sendOperation(new GetLastCommandResultOp(serial));
        switch (r.getStatus()) {
            case EXECUTING:
                handler.stillRunning();
                break;
            case INVOCATION_ERROR:
                handler.failure(r.getInvocationErrorDetails(), r.getFreeDeviceState(),
                        r.getRunMetrics());
                break;
            case INVOCATION_SUCCESS:
                handler.success(r.getRunMetrics());
                break;
            case NO_ACTIVE_COMMAND:
                handler.noActiveCommand();
                break;
            case NOT_ALLOCATED:
                handler.notAllocated();
                break;
            default:
                throw new RemoteException("unrecognized status " + r.getStatus().name());
        }
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public synchronized void close() {
        if (mSocket != null) {
            try {
                mSocket.close();
            } catch (IOException e) {
                Log.w(TAG, String.format("exception closing socket: %s", e.toString()));
            }
        }
        if (mWriter != null) {
            mWriter.close();
        }
    }
}
